groups/ntc/ntcq/ntcq_zerocopy.cpp

```diff
@@ -19,187 +19,453 @@
-: d_callbacks(allocator)
-, d_sendContext()
-, d_data_sp()
-, d_id(0)
-, d_allocator_p(allocator)
+bsl::ostream& ZeroCopyRange::print(bsl::ostream& stream,
+                                   int           level,
+                                   int           spacesPerLevel) const
 {
+    bslim::Printer printer(&stream, level, spacesPerLevel);
+    printer.start();
+
+    if (d_minCounter == d_maxCounter) {
+        printer.printValue("EMPTY");
+    }
+    else {
+        printer.printAttribute("min", d_minCounter);
+        printer.printAttribute("max", d_maxCounter);
+    }
+
+    printer.end();
+    return stream;
 }
 
-ZeroCopyEntry::ZeroCopyEntry(const ZeroCopyEntry& other,
-                             bslma::Allocator*    allocator)
-: d_callbacks(other.d_callbacks)
-, d_sendContext(other.d_sendContext)
-, d_data_sp(other.d_data_sp)
-, d_id(other.d_id)
-, d_allocator_p(allocator)
+void ZeroCopyEntry::match(const ntcq::ZeroCopyRange& complete)
 {
+    if (d_rangeSet.empty()) {
+        ntcq::ZeroCopyRange required = d_range;
+
+        ntcq::ZeroCopyRange intersection =
+            ntcq::ZeroCopyRange::intersect(required, complete);
+
+        if (intersection.empty()) {
+            return;
+        }
+
+        ntcq::ZeroCopyRange pending;
+        ntcq::ZeroCopyRange overflow;
+
+        ntcq::ZeroCopyRange::difference(
+            &pending, &overflow, required, intersection);
+
+        if (overflow.empty()) {
+            d_range = pending;
+        }
+        else {
+            d_rangeSet.push_back(pending);
+            d_rangeSet.push_back(overflow);
+            d_range.reset();
+        }
+
+        return;
+    }
+
+    RangeSet rangeSet(d_allocator_p);
+    rangeSet.swap(d_rangeSet);
+
+    RangeSet::const_iterator it = rangeSet.begin();
+    RangeSet::const_iterator et = rangeSet.end();
+
+    for (; it != et; ++it) {
+        ntcq::ZeroCopyRange required = *it;
+
+        ntcq::ZeroCopyRange intersection =
+            ntcq::ZeroCopyRange::intersect(required, complete);
+
+        if (intersection.empty()) {
+            d_rangeSet.push_back(required);
+            continue;
+        }
+
+        ntcq::ZeroCopyRange pending;
+        ntcq::ZeroCopyRange overflow;
+
+        ntcq::ZeroCopyRange::difference(
+            &pending, &overflow, required, intersection);
+
+        if (!pending.empty()) {
+            d_rangeSet.push_back(pending);
+        }
+
+        if (!overflow.empty()) {
+            d_rangeSet.push_back(overflow);
+        }
+    }
 }
 
-ZeroCopyEntry::~ZeroCopyEntry()
+bool ZeroCopyEntry::complete() const
 {
+    if (d_framed) {
+        if (d_rangeSet.empty()) {
+            return d_range.empty();
+        }
+        else {
+            RangeSet::const_iterator it = d_rangeSet.begin();
+            RangeSet::const_iterator et = d_rangeSet.end();
+
+            for (; it != et; ++it) {
+                const ntcq::ZeroCopyRange& range = *it;
+                if (!range.empty()) {
```

Comment:
I cannot see how there can be an empty range inside the range set. It looks like that if there is a non empty range set, then the entry is not complete. 
But I haven't yet read the test driver.

Reply:
This implementation is for robustness in potential future changes, or possible implementations. Given multiple elements in `d_rangeSet`, when we intersect/substract the completed range to calculate an empty set, its not clear to me its in our best interest to erase that element in the set (maybe causing us to shift and copy remaining elements).

In other words, this implementation allows `d_rangeSet` of `[0, 0), [ 3-6), [0,0), [ 9, 12), [0,0)` to be equivalent to `[ 3-6), [ 9, 12)`.

Reply:
Ok